tests/TestCase/ORM/Association/HasManyTest.php

```diff
@@ -693,4 +693,36 @@ public function testUnlinkWithEmptyArray()
         $this->assertSame(4, $this->author->find()->count(), 'Authors should still exist');
         $this->assertSame(3, $articles->find()->count(), 'Articles should still exist');
     }
+
+    /**
+     * Tests that link only uses a single database transaction
+     *
+     * @return void
+     */
+    public function testLinkUsesSingleTransaction()
+    {
+        $articles = TableRegistry::get('Articles');
+        $assoc = $this->author->hasMany('Articles', [
+            'sourceTable' => $this->author,
+            'targetTable' => $articles
+        ]);
+
+        // Ensure author in fixture has zero associated articles
+        $entity = $this->author->get(2, ['contain' => 'Articles']);
+        $initial = $entity->articles;
+        $this->assertCount(0, $initial);
+
+        // Ensure that after each model is saved, we are still within a transaction.
+        $listenerAfterSave = function ($e, $entity, $options) use ($articles) {
+            $debugInfo = $articles->connection()->__debugInfo();
```

Comment:
I completely agree.

The problem I faced was that each individual transaction was wrapped in a transactional() call, therefore even though it was multiple transactions, each call was still within a transaction. Therefore I had to measure transactionLevel.

Your question made me realise I can pass $options=[atomic=>false], then each individual save is not in it's own transaction, therefore connection()->inTransaction() will return the data I expect. I will amend my pull request now.